### DISCUSSION TITLE
Using autodispose family provider from long time provider

### DISCUSSION THREAD

**Discussion by balazsszekely** [rrousselGit/riverpod#2210] [Q&A]
Hi everyone, I've just started getting to know the Riverpod library and I really like it, but I've run into a problem that I can't seem to find a solution for anywhere.

I'm using a stream and I want to create another stream depending on the data coming in through the first stream.

Something like this:
```dart
final data1Provider = StreamProvider<int>((ref) {
  final data1Repo = ref.watch(data1RepoProvider);
  return data1Repo.dataStream;
});

final getData2Provider =
    StreamProvider.autoDispose.family<Data2?, String>((ref, data1Id) {
  final data2Repo = ref.watch(data2RepoProvider);
  return data2Repo.getData2Stream(data1Id);
});

final currentData2Provider = FutureProvider((ref) async {
  final data1 = await ref.watch(data1Provider.future);

  if (data1 == null) {
    return null;
  }

  final data2 = await ref.watch(getData2Provider(data1.id).future);
  return data2;
});
```

Now this code doesn't compile: https://docs-v2.riverpod.dev/docs/concepts/modifiers/auto_dispose#the-argument-type-autodisposeprovider-cant-be-assigned-to-the-parameter-type-alwaysaliveproviderbase

getData2Provider must be created using autoDispose modifier to avoid memory leak. But I never want to dispose currentData2Provider.

How should I solve this problem?

Thanks,
Balázs

**Comment by rrousselGit** [comment]
Since getData2Provider is autoDispose, then currentData2Provider _has_ to be autoDispose too.

That's not negotiable :)
Otherwise getData2Provider would never get disposed.

If you don't need getData2Provider to be autoDispose, you can disable it.

**Reply by balazsszekely** [reply]
Thanks for your fast reply. Then I will use the `autodispose` modifier and start listening to `currentData2Provider` in the app root. It is acceptable for me :)

As I played with it I created a more realistic sample: https://dartpad.dev/?id=ef06ab3ce0b822e6cc5db0575248e6e2, where `Data1` is `AuthState` and `Data2` is `UserData`.

In the previous code I wrote, I used `final data2 = await ref.watch(getData2Provider(data1.id).future);`, but it didn't work, because every time the data2Stream provided the second value, data2StreamProvider was disposed and started again, which results infinite loop. 

Is this the expected behaviour?

This can be reproduced in the DartPad linked above if we change the `authenticatedUserDataProvider` to `authenticatedUserDataProviderUsingFuture` in line 162.

My solution for this problem:
``` dart
final authenticatedUserDataProvider = Provider.autoDispose((ref) {
  return ref.watch(authStateProvider).when(
      data: (authState) {
        if (authState == null) {
          return const AsyncValue.data(null);
        }
        return ref.watch(getUserDataProvider(authState.userId));
      },
      error: ((error, stackTrace) => AsyncValue.error(error, stackTrace)),
      loading: () => const AsyncValue.loading());
});
```

I guess it can be prettier, but it works :)

Thanks again! :)
